File: src/js/main/RichJsonCache.js
export class RichJsonCache {
    addressMap = new WeakMap();
    nextAddress = 0;

    level = 0;
    stack = {};

    /**
     * Resolves a unique address for given object.
     * @param {object} object
     * @returns {String} The address
     */
    resolveAddress(object) {
        if (!this.addressMap.has(object)) {
            this.addressMap.set(object, String(this.nextAddress++));
            return String(this.nextAddress - 1);
        }
        return this.addressMap.get(object);
    }
}

File: src/java/main/RichJsonCache.java
import java.util.HashMap;
import java.util.Map;

public class RichJsonCache {
    public int level = 0;
    public String cloneAddress = null;
    public Map<String, String> inheritances = new HashMap<>();
    public Map<String, Object> stack = new HashMap<>();

    public String resolveAddress(Object obj) {
        if (obj == null) return "null";
        return String.valueOf(System.identityHashCode(obj));
    }
}